Slim/Schema.pm

```diff
@@ -862,32 +862,39 @@ sub _objForDbUrl {
 	my ($url) = @_;
 
 	if ($url =~ /^db:(\w+)\.(.+)/ ) {
-		my ($class, $values) = ($1, $2);
+		my ($class, $values) = ($1, "me.$2"); # need to qualify the base class because of possible duplicate column names in prefetched relations
 
 		my $query = {};
 		for my $term (split('&', $values)) {
 			if ($term =~ /(.*)=(.*)/) {
 				my $key = $1;
-				next if $key eq 'composer.name' || $key eq 'work.title' || $key eq 'track.grouping';
 				my $value = Slim::Utils::Misc::unescape($2);
 
 				if (utf8::is_utf8($value)) {
 					utf8::decode($value);
 					utf8::encode($value);
 				}
 
+				if ($key eq "track.grouping") {
+					$key = "tracks.grouping"; # track relation is wrongly named in Slim::Schema::Album but fixing it there breaks other things, including MAI.
```

Comment:
Can you elaborate? I believe we should fix this at the source, if possible. I'm happy to fix MAI if that's needed.

Reply:
When adding a favorite to the playlist, the `db:` URL keys are converted into class names (see `Slim::Control::Commands::_playlistXtracksCommand_parseDbItem`), so the URL must be` db:...&track.grouping=...` to match the class name Track.

But when drilling down into a favorite, the same URL key (first part before the `.`) must match the relation name in the base class (eg `Slim::Schema::Album`) for DBIx to generate the correct alias in its SQL (see `Slim::Schema::_objForDbUrl`). But the relation name there is `tracks`, not `track`.

So I think that for the clever` db:...` URL parsing to work in all cases, the names of relations defined in `Slim::Schema` classes need to match the referred-to classes, which is not the case in `Slim::Schema::Album` for tracks/Track (nor in `Slim::Schema::Work` because I blindly copied the `has_many` relation from the Album class when I created Work).

So, making this fix in `Slim::Schema::Album` (and `Slim::Schema::Work`) leads to these errors when displaying an album:
```
[24-04-21 13:04:59.3311] Slim::Menu::AlbumInfo::__ANON__ (157) Album menu item "fullalbum" failed: Can't locate object method "tracks" via package "Slim::Schema::Album" at /home/darrell/projects/slimserver/Slim/Menu/AlbumInfo.pm line 353.
[24-04-21 13:04:59.3318] Slim::Menu::AlbumInfo::__ANON__ (157) Album menu item "duration" failed: Can't locate object method "tracks" via package "Slim::Schema::Album" at /home/darrell/projects/slimserver/Slim/Schema/Album.pm line 378.
[24-04-21 13:04:59.3475] Slim::Menu::AlbumInfo::__ANON__ (157) Album menu item "moreartwork" failed: Can't locate object method "tracks" via package "Slim::Schema::Album" at /home/darrell/projects/slimserver/Cache/InstalledPlugins/Plugins/MusicArtistInfo/LocalArtwork.pm line 89.
[24-04-21 13:04:59.3478] Slim::Menu::AlbumInfo::__ANON__ (157) Album menu item "morefileinfo" failed: Can't locate object method "tracks" via package "Slim::Schema::Album" at /home/darrell/projects/slimserver/Cache/InstalledPlugins/Plugins/MusicArtistInfo/LocalFile.pm line 208.
```

Looks easy to fix, but I wonder what else this will break?

Reply:
Here is a bit of analysis. I only have Material. Qobuz and MAI (and the default built-in plugins) installed in this environment):

```
darrell@darrell-z570:~/projects/slimserver$ grep -rnI '>tracks'

Cache/InstalledPlugins/Plugins/MusicArtistInfo/LocalArtwork.pm:89:	return trackInfoHandler($client, undef, $album->tracks->first);
Cache/InstalledPlugins/Plugins/MusicArtistInfo/Importer.pm:234:				$track = $albumObj->tracks->first;
Cache/InstalledPlugins/Plugins/MusicArtistInfo/LocalFile.pm:208:	return trackInfoHandler($client, undef, $album->tracks->first);
Cache/InstalledPlugins/Plugins/MaterialSkin/Plugin.pm:1250:        my @tracks = $playlist->tracks;
CPAN/MP3/Cut/Gapless/Track.pm:52:    for my $track ( $cut->tracks ) {
CPAN/MP3/Cut/Gapless.pm:122:    for my $track ( $cue->tracks ) {
CPAN/MP3/Cut/Gapless.pm:155:    for my $track ( $cut->tracks ) {
Slim/Plugin/MusicMagic/Plugin.pm:721:		$mixSeed = $currentItem->tracks->next->path;
Slim/Plugin/MusicMagic/Plugin.pm:1288:			my $trackObj = $obj->tracks->next;
Slim/Schema/RemotePlaylist.pm:27:	$self->tracksRef(\@tracks);
Slim/Schema/RemotePlaylist.pm:33:	push @{$self->tracksRef()}, @$tracksRef;
Slim/Schema/RemotePlaylist.pm:38:	my $tracksRef = $self->tracksRef();
Slim/Schema/RemotePlaylist.pm:54:	for my $track ( $playlist->tracks ) {
Slim/Schema/Playlist.pm:144:	for my $track ( $playlist->tracks ) {
Slim/Schema/Album.pm:378:	foreach ($self->tracks) {
Slim/Player/Playlist.pm:1037:		[ $playlistObj->tracks ],
Slim/Control/Queries.pm:2633:		$iterator = $playlistObj->tracks($libraryId);
Slim/Control/Commands.pm:887:			$contents = [ map { $_->url } $playlist->tracks ];
Slim/Control/Commands.pm:1983:			$request->addResult( 'count', $playlist->tracks($library_id)->count() );
Slim/Control/Commands.pm:2177:	my @items   = $playlist->tracks;
Slim/Control/Commands.pm:2502:			my $playlistTrackIds = join(':', map { $_->id } $playlistObj->tracks);
Slim/Control/Commands.pm:2513:			[ $playlistObj->tracks ],
Slim/Control/Commands.pm:3405:			return $playlist->tracks($library_id);
Slim/Utils/Scanner/Remote.pm:1278:						for my $track ( $playlist->tracks ) {
Slim/Menu/AlbumInfo.pm:353:	my $totalAlbumTracks = $album->tracks->count;
Slim/Menu/PlaylistInfo.pm:197:			. Slim::Utils::Misc::delimitThousands($playlist->tracks($library_id)->count),

```

Reply:
Duplicating the relation in Slim::Schema::Album seems to work - still a bit hacky, though:

```
	$class->has_many('track'            => 'Slim::Schema::Track'            => 'album');
	$class->has_many('tracks'            => 'Slim::Schema::Track'            => 'album');

```

Reply:
Ok, I think I understand better now, thanks. If the duplication works, I'd prefer that. Please leave a note there for the reason why it's required.

Reply:
It is done.